enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,477 @@
+
+- Change `kubelet` configuration for OpenStack in the [MCO](https://github.com/openshift/machine-config-operator/blob/2a51e49b0835bbd3ac60baa685299e86777b064f/pkg/controller/template/render.go#L310-L357) to adopt external cloud providers.
+
+#### Examples for other cloud-controller-manager configurations
+
+- AWS: https://github.com/kubernetes/cloud-provider-aws/
+- Azure: https://github.com/kubernetes-sigs/cloud-provider-azure
+Azure separated the `CCM` and the "cloud-node-manager" (`CNM`) into two binaries: https://github.com/kubernetes-sigs/cloud-provider-azure/tree/master/cmd.
+- GCP: https://github.com/kubernetes/cloud-provider-gcp
+- vSphere: https://github.com/kubernetes/cloud-provider-vsphere
+vSphere introduced an additional service to expose internal API server: https://github.com/kubernetes/cloud-provider-vsphere/blob/master/manifests/controller-manager/vsphere-cloud-controller-manager-ds.yaml#L64
+
+### Risks and Mitigations
+
+#### Specific CCM doesn’t work properly on OpenShift
+
+The `CCMs` have not been tested on OCP yet. Unknown issues may arise.
+
+Severity: medium-high
+Likelihood: medium
+
+#### OpenShift components can't work with the CCM properly
+
+Since OpenShift imposes additional limitations compared to Kubernetes, some collisions are possible.
+So far there are no CCMs available in OpenShift, and there can be some problems with operability.
+
+Severity: medium-high
+Likelihood: low
+
+#### CSI driver is not available in OpenShift
+
+CCM doesn't support the in-tree `Volume` manager implementation. A CSI driver now becomes a mandatory requirement for every out-of-tree provider. Additionally the CSI should support data migration between the in-tree storage solution and the plugin provided storage interfaces.
+
+Severity: medium-high
+Likelihood: low
+
+### Test Plan
+
+- Make sure the `CCM` implementation for "any" cloud is up and running in a bootstrap phase before or shortly after the `kubelet` is running on any `Node`.
+- Ensure that transition between the in-tree and the out-of-tree will be handled by the established architecture with the new release payload, where all of the existing `Nodes` are operational during and after the upgrade time, and the bootstrap procedure for new ones is successful.
+- Ensure that upgrade from a release running on the in-tree provider only to the out-of-tree succeeds, and the downgrade is also supported.
+- Each provider should have a working CI implementation, which will assist in testing the provisioning and upgrades with out-of-tree support enabled.
+
+### Graduation Criteria
+
+#### Tech preview
+
+Operator handles OpenStack installation in 4.7.
```

Comment:
this surprises me. Does this mean we only switch to external for OpenShift as opt-in i.e. optionally?

Reply:
Mike suggests some optional environment variable for the installer to recognize it. Could you please elaborate, @Fedosin?